### PR TITLE
FOLSPRINGB-41: Add TenantService event methods for module overriding

### PR DESCRIPTION
**Pull request by ncovercash**
# [Jira FOLSPRINGB-41](https://issues.folio.org/projects/FOLSPRINGB/issues/FOLSPRINGB-41)

## Purpose

For some modules, additional work must be done in the event of tenant creation/deletion.  This can currently be done by extending TenantService in the module, however, this gets rather nasty due to the use of private fields/methods and requires copying a substantial amount of logic (including constructors) and could easily break down after a logic change in this module.

Furthermore, there is no feasible way of loading sample or reference data without completely hijacking the controller at the core levels of Spring (since multiple controllers may not map to the same requests).

## Inspiration

In my usage, I am rewriting a module from RMB/jsonb to Spring/relational.  Due to the complete overhaul of the database schema (not to mention the change in data structure itself to something more consistent/documented), Liquibase alone was not nearly enough.

With this, I can now add a simple `beforeLiquibaseUpdate` to check for any legacy schemas and extract their data.  Then, on `afterLiquibaseUpdate`, I can transform and save all this data into the new schema, ensuring a seamless migration.

## Approach

This adds a series of methods to `TenantService` to allow custom logic in/around the tenant creation/upgrade/deletion processes, encouraging the service to be extended and customized to fit any module's needs.

<details>
  <summary>Here is an excerpt from the updated README detailing the added methods.</summary>

  ## Custom `/_/tenant` Logic
  
  There are many cases where you may want to add custom logic to the
  [`/_/tenant` endpoint](https://s3.amazonaws.com/foliodocs/api/folio-spring-base/s/tenant.html),
  such as for loading sample data or performing more complex database migration.
  
  In order to do this, you can extend the `TenantService` within your module and override
  any of the methods listed below.
  
  ### `TenantService` Event Methods
  
  The following methods can be overridden by your module in order to add custom logic around events
  relating to tenant creation, updates, and deletion. All of these return `void`.
  
  Many of these accept a `TenantAttributes` parameter which can provide information about the
  previous module (`module_from`), the module being upgraded to (`module_to`), as well as any other
  `parameters` provided.
  
  :warning: Please note that methods with "update" in the name will be ran on updates _as well as_ when new
  tenants are created. Be especially careful with methods ran before Liquibase -- the database
  schema could potentially be in an unexpected state, particularly when a new tenant is created.
  
  | Visibility  | Signature                                 | Purpose                                                                                         |
  | ----------- | ----------------------------------------- | ----------------------------------------------------------------------------------------------- |
  | `public`    | `loadReferenceData()`                     | Load any reference data (requested with `loadReference=true` parameter)                         |
  | `public`    | `loadSampleData()`                        | Load any sample data (requested with `loadSample=true` parameter)                               |
  | `protected` | `beforeTenantUpdate(TenantAttributes)`    | Run custom logic before a tenant is created or updated                                          |
  | `protected` | `beforeLiquibaseUpdate(TenantAttributes)` | Run custom logic immediately before Liquibase updates are started (after `beforeTenantUpdate`)  |
  | `protected` | `afterLiquibaseUpdate(TenantAttributes)`  | Run custom logic immediately before Liquibase updates are finished (before `afterTenantUpdate`) |
  | `protected` | `afterTenantUpdate(TenantAttributes)`     | Run custom logic after all update jobs are completed                                            |
  | `protected` | `beforeTenantDeletion(TenantAttributes)`  | Run custom logic before a tenant is deleted/purged                                              |
  | `protected` | `afterTenantDeletion(TenantAttributes)`   | Run custom logic after a tenant is deleted/purged (the schema will no longer exist)             |
  
  ### `TenantService` Methods and Fields
  
  There are two methods that may be of use in your custom logic:
  
  - `boolean tenantExists()` which will check if the database schema for this tenant exists (this
    says nothing about if it is up to date)
  - `String getSchemaName()` will construct and return the name of the schema corresponding to the
    module and tenant
  
  These fields will also be provided:
  
  - `JdbcTemplate jdbcTemplate`, for running Postgres queries directly
  - `FolioExecutionContext context`, for getting information about the module
  - `FolioSpringLiquibase folioSpringLiquibase`, for interacting with Liquibase directly (this
    extends `SpringLiquibase` and may be `null` if Liquibase is not enabled!)
  
  ### Event Order
  
  The [events](#tenantservice-event-methods) will be called in the following order:
  
  #### Upon Creation
  
  1. `beforeTenantUpdate`
  2. If Liquibase is enabled:
     1. `beforeLiquibaseUpdate`
     2. _Internal logic to apply Liquibase changes_
     3. `afterLiquibaseUpdate`
  3. `afterTenantUpdate`
  4. `loadReferenceData`, if applicable
  5. `loadSampleData`, if applicable
  
  #### Upon Deletion
  
  1. `beforeTenantDeletion`
  2. _Internal logic to drop the schema_
  3. `afterTenantDeletion`
  
  ### Sample
  
  Overriding these methods to add your own custom logic is quite straightforward. Here is an example
  of how to override these in your very own `@Service`:
  
  ```java
  package org.folio.yourmodule.service;
  
  import javax.annotation.Priority;
  import org.folio.spring.service.TenantService;
  import org.folio.tenant.domain.dto.TenantAttributes;
  import org.folio.yourmodule.SuperCoolDataRepository;
  import org.springframework.beans.factory.annotation.Autowired;
  import org.springframework.stereotype.Service;
  
  @Service
  @Priority(-10) // required to ensure CustomTenantService will be loaded instead of TenantService
  public class CustomTenantService extends TenantService {

    @Autowired
    protected final SuperCoolDataRepository repository;
  
    /**
     * Load reference data
     */
    @Override
    protected void loadReferenceData() {
      repository.loadReferenceData();
    }
  
    /**
     * Add our custom initial data
     */
    @Override
    protected void beforeTenantUpdate(TenantAttributes attributes) {
      // some custom logic for potentially migrating data
    }
  }
  ```

</details>

## TODOS and Open Questions

- [ ] I'm not sure that any unit tests can really be added here since no internal logic was really changed.
- [ ] I see that v3.0.1 is currently in the process of being released -- please let me know if I need to rebase anything due to merge errors!
- [ ] Have feedback for these method names, places, etc?  Let me know!  I [posted in Slack](https://folio-project.slack.com/archives/C01HUKJCMC7/p1642713846000900) about it, however, no one replied, so I decided to go with my own intuition on these.
  - I feel like the difference in liquibase update methods and non-Liquibase methods is negligible, however, am not sure on this.  Personally, I think more choice is better and this allows special logic for people that want to ensure Liquibase is working properly (by overriding the `*LiquibaseUpdate` methods) whereas other modules, who may not even use Liquibase, can use the more generic `*TenantUpdate` methods.

## Other Improvements

As I was working, I did the following other small things of potential note:

- Bump the copyright year to 2022
- Fix a few README formatting/spelling issues
- Add some Javadoc to most `TenantService` methods
- Rename `TenantService::createTenant` to `TenantService::createOrUpdateTenant` to better describe its usage

## Learning

[`@Priority` documentation](https://docs.oracle.com/javaee/7/api/javax/annotation/Priority.html)

[Spring's use of `@Priority` and advantages over `@Primary`](https://github.com/spring-projects/spring-framework/issues/15179)

### CODE REVIEW

File: src/main/java/org/folio/spring/controller/TenantController.java [hunk 1]
```diff
@@ -34,40 +34,24 @@
...
+      tenantService.deleteTenant(tenantAttributes);
     } else {
       log.info("Upgrading tenant...");
-      upgradeTenant();
+      tenantService.createOrUpdateTenant(tenantAttributes);
       var parameters = tenantAttributes.getParameters();
       var loadReferenceParam = getParameterValue(LOAD_REFERENCE_PARAM, parameters);
       if (loadReferenceParam.isPresent() && loadReferenceParam.get().equals("true")) {
-        loadReferenceData();
+        tenantService.loadReferenceData();
```
Comment (dmtkachenko) [inline src/main/java/org/folio/spring/controller/TenantController.java:44]: Agree, these should be moved to the service

File: README.md [hunk 2]
```diff
@@ -42,54 +64,176 @@ public interface JpaCqlRepository<T, ID> extends JpaRepository<T, ID> {
...
+
+import javax.annotation.Priority;
+import org.folio.spring.service.TenantService;
+import org.folio.tenant.domain.dto.TenantAttributes;
+import org.folio.yourmodule.SuperCoolDataRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+@Priority(-10) // required to ensure CustomTenantService will be loaded instead of TenantService
```
Comment (dmtkachenko) [inline README.md:218]: Just a note:
I'd probably got with `@Primary` annotation instead. These shouldn't be more than one implementation of TenantService, so using `Priority` or `Order` that assume several options in general is less appropriate here in my opinion, but Priority will also work.

**Comment by dmtkachenko** [comment]
Look good in general.
Please rebase to the latest version from master

File: src/main/java/org/folio/spring/controller/TenantController.java (hunk 1 above)
Comment (ncovercash) [reply src/main/java/org/folio/spring/controller/TenantController.java:44]: :+1:

**Review by dmtkachenko — APPROVED** [review]

**Comment by ncovercash** [comment]
@psmagin done; anything else that I need to do for this?

**Comment by ncovercash** [comment]
Is this good for merge? @psmagin @dmtkachenko

**Comment by ncovercash** [comment]
@dmtkachenko sorry, thought I fixed that.  Let me look into that.

**Comment by ncovercash** [comment]
Ah, didn't see the revert and was very confused.  @dmtkachenko all should be fixed!

**Pull request merged**